File: scenarios/code-facts/api/internal/testutil/modeltest/matrix.go
// Package modeltest contains test-only helpers for explicit workflow models.
package modeltest

import (
	"fmt"
	"strings"
)

// TestingT is the subset of *testing.T used by this package.
type TestingT interface {
	Helper()
	Fatalf(format string, args ...any)
}

// Transition applies event to state and returns the next state.
type Transition[S comparable, E comparable] func(state S, event E) (S, error)

// MatrixRow describes the expected result for one state/event pair.
type MatrixRow[S comparable, E comparable] struct {
	Name    string
	From    S
	Event   E
	To      S
	WantErr bool
}

type pair[S comparable, E comparable] struct {
	state S
	event E
}

// ValidateTransitionMatrix checks that rows cover every state/event pair
// exactly once and match the supplied production transition function.
func ValidateTransitionMatrix[S comparable, E comparable](
	statuses []S,
	events []E,
	rows []MatrixRow[S, E],
	transition Transition[S, E],
) []error {
	var errs []error
	if len(statuses) == 0 {
		errs = append(errs, fmt.Errorf("statuses must not be empty"))
	}
	if len(events) == 0 {
		errs = append(errs, fmt.Errorf("events must not be empty"))
	}
	if transition == nil {
		errs = append(errs, fmt.Errorf("transition must not be nil"))
	}

	knownStatuses := make(map[S]struct{}, len(statuses))
	for _, status := range statuses {
		if _, ok := knownStatuses[status]; ok {
			errs = append(errs, fmt.Errorf("duplicate status %v", status))
			continue
		}
		knownStatuses[status] = struct{}{}
	}

	knownEvents := make(map[E]struct{}, len(events))
	for _, event := range events {
		if _, ok := knownEvents[event]; ok {
			errs = append(errs, fmt.Errorf("duplicate event %v", event))
			continue
		}
		knownEvents[event] = struct{}{}
	}

	seen := make(map[pair[S, E]]string, len(rows))
	for i, row := range rows {
		label := row.Name
		if label == "" {
			label = fmt.Sprintf("row %d", i)
		}
		if _, ok := knownStatuses[row.From]; !ok {
			errs = append(errs, fmt.Errorf("%s: unknown from status %v", label, row.From))
		}
		if _, ok := knownStatuses[row.To]; !ok {
			errs = append(errs, fmt.Errorf("%s: unknown to status %v", label, row.To))
		}
		if _, ok := knownEvents[row.Event]; !ok {
			errs = append(errs, fmt.Errorf("%s: unknown event %v", label, row.Event))
		}

		key := pair[S, E]{state: row.From, event: row.Event}
		if first, ok := seen[key]; ok {
			errs = append(errs, fmt.Errorf("%s: duplicate pair %v/%v already covered by %s", label, row.From, row.Event, first))
			continue
		}
		seen[key] = label
	}

	for _, status := range statuses {
		for _, event := range events {
			if _, ok := seen[pair[S, E]{state: status, event: event}]; !ok {
				errs = append(errs, fmt.Errorf("missing pair %v/%v", status, event))
			}
		}
	}
	if len(errs) > 0 || transition == nil {
		return errs
	}

	for i, row := range rows {
		label := row.Name
		if label == "" {
			label = fmt.Sprintf("row %d", i)
		}
		got, err := transition(row.From, row.Event)
		if row.WantErr {
			if err == nil {
				errs = append(errs, fmt.Errorf("%s: expected error, got nil", label))
			}
		} else if err != nil {
			errs = append(errs, fmt.Errorf("%s: unexpected error: %v", label, err))
		}
		if got != row.To {
			errs = append(errs, fmt.Errorf("%s: got status %v, want %v", label, got, row.To))
		}
	}
	return errs
}

// AssertTransitionMatrix fails t when ValidateTransitionMatrix finds drift.
func AssertTransitionMatrix[S comparable, E comparable](
	t TestingT,
	statuses []S,
	events []E,
	rows []MatrixRow[S, E],
	transition Transition[S, E],
) {
	t.Helper()
	if errs := ValidateTransitionMatrix(statuses, events, rows, transition); len(errs) > 0 {
		t.Fatalf("transition matrix mismatch:\n%s", formatErrors(errs))
	}
}

func formatErrors(errs []error) string {
	lines := make([]string, 0, len(errs))
	for _, err := range errs {
		lines = append(lines, "  - "+err.Error())
	}
	return strings.Join(lines, "\n")
}

File: scenarios/react-component-library/ui/src/test-utils/modeltest/matrix.ts
export type WorkflowTransition<State extends PropertyKey, Event extends PropertyKey> = (
  state: State,
  event: Event,
) => State;

export interface MatrixRow<State extends PropertyKey, Event extends PropertyKey> {
  readonly name?: string;
  readonly from: State;
  readonly event: Event;
  readonly to: State;
  readonly wantError?: boolean;
}

const pairKey = (state: PropertyKey, event: PropertyKey) =>
  `${String(state)}\u0000${String(event)}`;

export const validateTransitionMatrix = <State extends PropertyKey, Event extends PropertyKey>(
  states: readonly State[],
  events: readonly Event[],
  rows: readonly MatrixRow<State, Event>[],
  transition: WorkflowTransition<State, Event>,
): string[] => {
  const errors: string[] = [];
  if (states.length === 0) {
    errors.push("states must not be empty");
  }
  if (events.length === 0) {
    errors.push("events must not be empty");
  }

  const knownStates = new Set<State>();
  states.forEach((state) => {
    if (knownStates.has(state)) {
      errors.push(`duplicate state ${String(state)}`);
      return;
    }
    knownStates.add(state);
  });

  const knownEvents = new Set<Event>();
  events.forEach((event) => {
    if (knownEvents.has(event)) {
      errors.push(`duplicate event ${String(event)}`);
      return;
    }
    knownEvents.add(event);
  });

  const seen = new Map<string, string>();
  rows.forEach((row, index) => {
    const label = row.name ?? `row ${index}`;
    if (!knownStates.has(row.from)) {
      errors.push(`${label}: unknown from state ${String(row.from)}`);
    }
    if (!knownStates.has(row.to)) {
      errors.push(`${label}: unknown to state ${String(row.to)}`);
    }
    if (!knownEvents.has(row.event)) {
      errors.push(`${label}: unknown event ${String(row.event)}`);
    }

    const key = pairKey(row.from, row.event);
    const first = seen.get(key);
    if (first) {
      errors.push(
        `${label}: duplicate pair ${String(row.from)}/${String(row.event)} already covered by ${first}`,
      );
      return;
    }
    seen.set(key, label);
  });

  states.forEach((state) => {
    events.forEach((event) => {
      if (!seen.has(pairKey(state, event))) {
        errors.push(`missing pair ${String(state)}/${String(event)}`);
      }
    });
  });

  if (errors.length > 0) {
    return errors;
  }

  rows.forEach((row, index) => {
    const label = row.name ?? `row ${index}`;
    try {
      const got = transition(row.from, row.event);
      if (row.wantError) {
        errors.push(`${label}: expected error, got success`);
      }
      if (got !== row.to) {
        errors.push(`${label}: got state ${String(got)}, want ${String(row.to)}`);
      }
    } catch (error) {
      if (!row.wantError) {
        errors.push(
          `${label}: unexpected error: ${error instanceof Error ? error.message : String(error)}`,
        );
      }
      if (row.from !== row.to) {
        errors.push(
          `${label}: error rows must keep state unchanged; got ${String(row.from)}, want ${String(row.to)}`,
        );
      }
    }
  });

  return errors;
};

export const assertTransitionMatrix = <State extends PropertyKey, Event extends PropertyKey>(
  states: readonly State[],
  events: readonly Event[],
  rows: readonly MatrixRow<State, Event>[],
  transition: WorkflowTransition<State, Event>,
): void => {
  const errors = validateTransitionMatrix(states, events, rows, transition);
  if (errors.length > 0) {
    throw new Error(`transition matrix mismatch:\n${errors.map((err) => `  - ${err}`).join("\n")}`);
  }
};
